service/frontend/http_api_server.go

```diff
@@ -0,0 +1,438 @@
+func (h *HTTPAPIServer) GracefulStop() {
+	// We try a graceful stop for the amount of time we can drain, then we do a
+	// hard stop
+	shutdownCtx, _ := context.WithTimeout(context.Background(), h.shutdownDrainTime)
+	// We intentionally ignore this error, we're gonna stop at this point no
+	// matter what. This closes the listener too.
+	_ = h.server.Shutdown(shutdownCtx)
+	_ = h.server.Close()
+	close(h.stopped)
+}
+
+func (h *HTTPAPIServer) serveHTTP(w http.ResponseWriter, r *http.Request) {
+	// Limit the request body to max gRPC size. This is hardcoded to 4MB at the
+	// moment using gRPC's default at
+	// https://github.com/grpc/grpc-go/blob/0673105ebcb956e8bf50b96e28209ab7845a65ad/server.go#L58.
+	// Max header bytes is defaulted by Go to 1MB.
+	r.Body = http.MaxBytesReader(w, r.Body, 1024*1024*4)
+
+	h.logger.Debug(
+		"HTTP API call",
+		tag.NewStringTag("http-method", r.Method),
+		tag.NewAnyTag("http-url", r.URL),
+	)
+
+	// Need to change the accept header based on whether pretty and/or
+	// noPayloadShorthand are present
+	var acceptHeaderSuffix string
+	if _, ok := r.URL.Query()["pretty"]; ok {
+		acceptHeaderSuffix += "+pretty"
+	}
+	if _, ok := r.URL.Query()["noPayloadShorthand"]; ok {
+		acceptHeaderSuffix += "+no-payload-shorthand"
+	}
+	if acceptHeaderSuffix != "" {
+		r.Header.Set("Accept", "application/json"+acceptHeaderSuffix)
+	}
+
+	// Put the TLS info on the peer context
+	if r.TLS != nil {
+		var addr net.Addr
+		if conn, _ := r.Context().Value(httpRemoteAddrContextKey{}).(net.Conn); conn != nil {
+			addr = conn.RemoteAddr()
+		}
+		r = r.WithContext(peer.NewContext(r.Context(), &peer.Peer{
+			Addr: addr,
+			AuthInfo: credentials.TLSInfo{
+				State:          *r.TLS,
+				CommonAuthInfo: credentials.CommonAuthInfo{SecurityLevel: credentials.PrivacyAndIntegrity},
+			},
+		}))
+	}
+
+	// Call gRPC gateway mux
+	h.serveMux.ServeHTTP(w, r)
+}
+
+func (h *HTTPAPIServer) errorHandler(
+	ctx context.Context,
+	mux *runtime.ServeMux,
+	marshaler runtime.Marshaler,
+	w http.ResponseWriter,
+	r *http.Request,
+	err error,
+) {
+	// Convert the error using serviceerror. The result does not conform to Google
+	// gRPC status directly (it conforms to gogo gRPC status), but Err() does
+	// based on internal code reading. However, Err() uses Google proto Any
+	// which our marshaler is not expecting. So instead we are embedding similar
+	// logic to runtime.DefaultHTTPProtoErrorHandler in here but with gogo
+	// support. We don't implement custom content type marshaler or trailers at
+	// this time.
+
+	s := serviceerror.ToStatus(err)
+	w.Header().Set("Content-Type", marshaler.ContentType())
+
+	buf, merr := marshaler.Marshal(s.Proto())
+	if merr != nil {
+		h.logger.Warn("Failed to marshal error message", tag.Error(merr))
+		w.WriteHeader(http.StatusInternalServerError)
+		_, _ = w.Write([]byte(`{"code": 13, "message": "failed to marshal error message"}`))
```

Comment:
should set content-type to json here? or is `marshaler.ContentType()` always json?

Reply:
`marshaler.ContentType()` is always JSON at this time, but I'll add it explicitly just in case. I expect this code will never be hit.

Reply:
Fixed